src/asn_prim.ml

```diff
@@ -84,60 +84,51 @@ module Null : Prim with type t = unit = struct
   let random () = ()
 end
 
-module Integer : Prim with type t = Z.t = struct
+module Integer : Prim_s with type t = Cstruct.t = struct
 
-  type t = Z.t
-
-  let of_int8 x = Z.of_int (if x >= 0x80 then x - 0x100 else x) [@@inline]
-  let of_int16 x = Z.of_int (if x >= 0x8000 then x - 0x10000 else x) [@@inline]
+  type t = Cstruct.t
 
+  (* The only check for an integer is that if the first byte is 0, the second
+     should be >= 0x80 *)
   let of_cstruct cs =
-    let open Cstruct in
-
-    let rec go acc i = function
-      n when n >= 8 ->
-        let w = Z.of_int64 (BE.get_uint64 cs i) in
-        go Z.((acc lsl 64) lor (extract w 0 64)) (i + 8) (n - 8)
-    | n when n >= 4 ->
-        let w = Z.of_int32 (BE.get_uint32 cs i) in
-        go Z.((acc lsl 32) lor (extract w 0 32)) (i + 4) (n - 4)
-    | n when n >= 2 ->
-        go Z.((acc lsl 16) lor ~$(BE.get_uint16 cs i)) (i + 2) (n - 2)
-    | 1 -> Z.((acc lsl 8) lor ~$(get_uint8 cs i))
-    | _ -> acc in
     match cs.Cstruct.len with
-      0 -> parse_error "INTEGER: length 0"
-    | 1 -> of_int8 (get_uint8 cs 0)
-    | n ->
-        let w0 = BE.get_uint16 cs 0 in
+    | 0 -> parse_error "INTEGER: length 0"
+    | 1 -> cs
+    | _ ->
+        let w0 = Cstruct.BE.get_uint16 cs 0 in
         match w0 land 0xff80 with
-          0x0000 | 0xff80 -> parse_error "INTEGER: redundant form"
-        | _ -> go (of_int16 w0) 2 (n - 2)
-
-  let last8 z = Z.(extract z 0 8 |> to_int)
-
-  let to_writer n =
-    let sz  = Z.size n * 8 + 1 in
-    let sz1 = sz - 1 in
-    let cs  = Cstruct.create sz in
+        | 0x0000 | 0xff80 -> parse_error "INTEGER: redundant form"
+        | _ -> cs
 
-    let rec write i n =
-      if n = Z.(~$(-1)) || n = Z.zero then i else
-        ( Cstruct.set_uint8 cs i (last8 n) ;
-          write (pred i) Z.(n asr 8) ) in
+  let to_writer = Writer.of_cstruct
 
-    let (bad_b0, padding) =
-      if n >= Z.zero then ((<=) 0x80, 0x00)
-      else ((>) 0x80, 0xff) in
-    let off =
-      let i = write sz1 n in
-      if i = sz1 || bad_b0 (Cstruct.get_uint8 cs (succ i)) then
-        ( Cstruct.set_uint8 cs i padding ; i )
-      else succ i in
-    Writer.of_cstruct Cstruct.(sub cs off (sz - off))
+  (* we actually produce integers that fit into an int *)
+  let random ?size () =
+    let rec one () =
+      let cs =
+        random_string ?size ~chars:(0, 256) () |> Cstruct.of_string
+      in
+      match Cstruct.length cs with
+      | 0 -> one ()
+      | 1 -> cs
+      | x ->
+        if (Sys.word_size = 32 && x <= 4) || (Sys.word_size = 64 && x <= 8) then
+          match Cstruct.BE.get_uint16 cs 0 land 0xff80 with
+          | 0x0000 | 0xff80 -> one ()
+          | _ ->
+            (* OCaml integer are only 31 / 63 bit *)
+            if Cstruct.get_uint8 cs 0 land 0xc0 <> 0 then
+              one ()
+            else
+              cs
```

Comment:
This I struggle a bit with. It checks if any of two topmost bits are set, and rejects it if any of them are.

Reply:
Testing some values this seems to work. Maybe I just need to look at it again with a fresh brain :-)

Reply:
The reason is: `Printf.printf "%X" max_int` (= 0x3FFFFFFFFFFFFFFF) and `Printf.printf "%X" min_int` (= 0x4000000000000000).

This means if the highest bits is set, we can't encode/decode it. And we should only use 0x80, not 0xc0 - well spotted.

Reply:
commented too early, but there's an issue with numbers at the edge. 0x4d1bee5c6072de29 (in OCaml -3667075391040594391) is decoded as Int64 5556296645814181417L), and then rejected.

Reply:
ok, so I pushed a fix. The numbers we can use are -- if of size 4 (8 for 64 bit) the first byte is <= 0x3F.

Since, in ASN.1, the 0x4000 0000 0000 0000 refers to 4611686018427387904, but in OCaml its value is -4611686018427387904.